### PR TITLE
improve title generation

### LINKED ISSUE #55: Note renaming from within app
Yesterday I installed this app and created some test notes. What was very disappointing to me is that one cannot rename a note from inside the app, but has to do it in the Notes folder. This is not very intuitive.

### PR DESCRIPTION
**Pull request by korelstar**
I've improved the generation of the note's title respective filename. Now, much more markdown markups as well as white spaces are removed. The approach is based on a similar implementation in the Android client: stefan-niedermann/nextcloud-notes#143

This will fix #55.

### CODE REVIEW

File: service/notesservice.php [hunk 1]
```diff
@@ -196,6 +181,33 @@ public function delete ($id, $userId) {
...
 
+    private function getSafeTitleFromContent($content) {
+        // prepare content: remove markdown characters and empty spaces
+        $content = preg_replace("/^\s*[*+-]\s+/m", "", $content); // list item
+        $content = preg_replace("/^#+\s+(.*?)\s*#*$/m", "$1", $content); // headline
+        $content = preg_replace("/^(?:=*|-*)$/m", "", $content); // separate line for headline
+        $content = preg_replace("/(\*+|_+)(.*?)\\1/m", "$2", $content); // emphasis
+        $content = preg_replace("/^\s+/m", "", $content); // whitespace at beginning
+        $content = preg_replace("/\s+$/m", "", $content); // whitespace at end
+        $content = trim($content);
```
Comment (Henni) [inline service/notesservice.php]: Trimming should remove whitespace both from the beginning and the end of the string. So why use the two `preg_replace` above?

File: service/notesservice.php [hunk 2]
```diff
@@ -196,6 +181,33 @@ public function delete ($id, $userId) {
...
+
+        // generate content from the first line of the title
+        $splitContent = preg_split("/\R/", $content, 2);
+        $title = $splitContent[0];
```
Comment (Henni) [inline service/notesservice.php]: how about moving this part up and just doing all the `preg_replace` stuff on the first line?

File: service/notesservice.php [hunk 3]
```diff
@@ -196,6 +181,33 @@ public function delete ($id, $userId) {
         $file->delete();
     }
 
+    private function getSafeTitleFromContent($content) {
+        // prepare content: remove markdown characters and empty spaces
+        $content = preg_replace("/^\s*[*+-]\s+/m", "", $content); // list item
+        $content = preg_replace("/^#+\s+(.*?)\s*#*$/m", "$1", $content); // headline
+        $content = preg_replace("/^(?:=*|-*)$/m", "", $content); // separate line for headline
+        $content = preg_replace("/(\*+|_+)(.*?)\\1/m", "$2", $content); // emphasis
```
Comment (Henni) [inline service/notesservice.php]: RegEx magic ✨
I still have to thoroughly test this stuff.

File: service/notesservice.php (hunk 2 above)
Comment (korelstar) [inline service/notesservice.php]: The advantage of doing the replacements first is that removing markups can produce an empty line which would result in an empty title. For example, if the first line consists of some characters that form a markdown code that is removed by the regexps, only some whitespaces are left. I wanted to prevent this, since an empty title is useless.

File: service/notesservice.php (hunk 1 above)
Comment (korelstar) [inline service/notesservice.php]: Yes, I can do that. I took the separate trimming from the Android app, since we had to remove markups for the whole content (the list of notes shows a small snippet of the note, beside the title).

File: service/notesservice.php (hunk 3 above)
Comment (Henni) [inline service/notesservice.php]: @privatereese do you want to review this?

File: service/notesservice.php [hunk 4]
```diff
@@ -196,6 +181,31 @@ public function delete ($id, $userId) {
         $file->delete();
     }
 
+    private function getSafeTitleFromContent($content) {
+        // prepare content: remove markdown characters and empty spaces
+        $content = preg_replace("/^\s*[*+-]\s+/m", "", $content); // list item
+        $content = preg_replace("/^#+\s+(.*?)\s*#*$/m", "$1", $content); // headline
```
Comment (Henni) [inline service/notesservice.php]: I wasn't aware of closing heading markup. I believe this currently leads to issues in the web editor. But this is unrelated to this PR.

File: service/notesservice.php [hunk 5]
```diff
@@ -196,6 +181,31 @@ public function delete ($id, $userId) {
...
+        $content = preg_replace("/^(?:=*|-*)$/m", "", $content); // separate line for headline
```
Comment (Henni) [inline service/notesservice.php]: Why do you explicitly use a non-capturing group? Does this have any advantages as opposed to `/^(=*|-*)$/m`.

Also shouldn't we write this as: `/^(=+|-+)$/m`? Otherwise empty lines match. Or was this intended?

Comment (korelstar) [inline service/notesservice.php]: I think, you're right. It's a long time ago, when I wrote those regular expressions. Therefore, I actually don't remember why I chose that variant. I will test your suggestion in detail and change it if it works.

Comment (korelstar) [inline service/notesservice.php]: @Henni , you were right. I've change the expression. Ready for merge?
